output/cloud/output_test.go

```diff
@@ -127,3 +132,177 @@ func TestOutputCreateTestWithConfigOverwrite(t *testing.T) {
 
 	require.NoError(t, out.StopWithTestError(nil))
 }
+
+func TestOutputStartVersionError(t *testing.T) {
+	t.Parallel()
+	o, err := newOutput(output.Params{
+		Logger: testutils.NewLogger(t),
+		ScriptOptions: lib.Options{
+			Duration:   types.NullDurationFrom(1 * time.Second),
+			SystemTags: &metrics.DefaultSystemTagSet,
+		},
+		Environment: map[string]string{
+			"K6_CLOUD_API_VERSION": "3",
+		},
+		ScriptPath: &url.URL{Path: "/script.js"},
+	})
+	require.NoError(t, err)
+
+	o.referenceID = "123"
+	err = o.startVersionedOutput()
+	require.ErrorContains(t, err, "v3 is an unexpected version")
+}
+
+func TestOutputStartWithReferenceID(t *testing.T) {
+	t.Parallel()
+
+	handler := func(w http.ResponseWriter, r *http.Request) {
+		// no calls are expected to the cloud service when
+		// the reference ID is passed
+		t.Error("got unexpected call")
+	}
+	ts := httptest.NewServer(http.HandlerFunc(handler))
+	defer ts.Close()
+
+	out, err := newOutput(output.Params{
+		Logger: testutils.NewLogger(t),
+		Environment: map[string]string{
+			"K6_CLOUD_HOST":        ts.URL,
+			"K6_CLOUD_PUSH_REF_ID": "my-passed-id",
+		},
+		ScriptOptions: lib.Options{
+			SystemTags: &metrics.DefaultSystemTagSet,
+		},
+		ScriptPath: &url.URL{Path: "/script.js"},
+	})
+	require.NoError(t, err)
+	require.NoError(t, out.Start())
+	require.NoError(t, out.Stop())
+}
+
+func TestCloudOutputDescription(t *testing.T) {
+	t.Parallel()
+
+	t.Run("WithTestRunDetails", func(t *testing.T) {
+		t.Parallel()
+		o := Output{referenceID: "74"}
+		o.config.TestRunDetails = null.StringFrom("my-custom-string")
+		assert.Equal(t, "cloud (my-custom-string)", o.Description())
+	})
+	t.Run("WithWebAppURL", func(t *testing.T) {
+		t.Parallel()
+		o := Output{referenceID: "74"}
+		o.config.WebAppURL = null.StringFrom("mywebappurl.com")
+		assert.Equal(t, "cloud (mywebappurl.com/runs/74)", o.Description())
+	})
+}
+
+func TestOutputStopWithTestError(t *testing.T) {
+	t.Parallel()
+
+	done := make(chan struct{}, 1)
```

Comment:
Why is the chan buffered? Could we produce more than one call? :thinking:

Reply:
Otherwise, it deadlocks on `StopWithTestError` waiting for the server to send the response.

My original idea was to have this test with the cloudapi.Client mocked so we don't need to have the test server and concurrency here but unfortunately, the v1 requires it so we can't just add an interface to mock it. When we will drop the v1 we may consider to do it to restrict the scope. Ideally, splitting between two interfaces one for the test life cycle management and one for metrics flush.

Reply:
Not sure if I'm getting why it could deadlock :thinking: 

Like I assume that the HTTP calling code is async. That's why you used a `chan` to ensure that call happened. If so, you don't need to send something to the chan, you could just close the chan once the handler responds.

If like you're saying, the StopWithTestError could lead to a deadlock. The code seems to me synchronous, and you probably don't need any chan at all :thinking: